### ISSUE TITLE
Support `autocomplete` on Input elements

### ISSUE THREAD

**Issue by TejasQ** [contiamo/operational-ui#725]
[Chrome supports and advocates](https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands) allowing these props on input elements. Operational UI's interfaces should allow them.

**Comment by adeelibr** [comment]
Can i work on this? 

As a suggestion, I have been working on React Material UI & what they do is they don't enforce `autoComplete` property. It is optional. The user should be exposed for `autoComplete` prop, it's type should be `string` and it the user passes that prop to the `Input` component it get's assigned to it otherwise it defaults to false, since it is `undefined`. Your thoughts? :)

**Comment by adeelibr** [comment]
On checking the `Input` component, `autoComplete` prop is already available. 😅 

Typings are already added

https://github.com/contiamo/operational-ui/blob/67ab23ed89f16e8e92286d0bfeba5cb7b713ed2a/src/Input/Input.tsx#L29

And then used here like

https://github.com/contiamo/operational-ui/blob/67ab23ed89f16e8e92286d0bfeba5cb7b713ed2a/src/Input/Input.tsx#L238

https://github.com/contiamo/operational-ui/blob/67ab23ed89f16e8e92286d0bfeba5cb7b713ed2a/src/Input/Input.tsx#L250


Although while looking at the code, I think `Input` should be self contained, while the component that take `Input` along with `label` should be a separate component. What do you think?

**Comment by TejasQ** [comment]
I think @peterszerzo would have amazing things to say on this.

**Comment by adeelibr** [comment]
So as a reference again from React Material UI, Is that they have a Input component which is just `Input` and then there is a `TextField` component which is composed on other components like `Input` . `InputLabel`, `FormControl`, etc..

In order to maintain compos ability I think every component should have a separate module. Like what @peterszerzo said https://github.com/contiamo/operational-ui/issues/726#issuecomment-422393321

**Comment by peterszerzo** [comment]
@adeelibr the reason we export `Input` component that also renders a label is encourage labeled inputs, which the user would have to compose themselves if we were to export these components separately.

Arguably this makes things a bit more complicated since it is not crystal clear from the outside whether extra HTML props would wind up on a `label` element or an `input` one, but I think it's worth handling this on a case-by-case basis (`autocomplete` doesn't make sense on a label, therefore it goes on input), to hang on to the public API convenience.

**Comment by adeelibr** [comment]
@peterszerzo So a great reference to support my argument is React Material UI `<TextField />` component https://material-ui.com/demos/text-fields/

This component is built on top of another component `Input`, where `<TextField />` is just composed of other components and exposed to the user. Like in your case. Does this make sense?

**Comment by peterszerzo** [comment]
Can you outline the benefits of changing from the current setup/public API to a Material UI-like one?

**Comment by adeelibr** [comment]
Yes, having an isolated `Input` component, gives as more freedom to build variations with a basic input component

- one variation can be of simple `Input` component nothing else
- another variation can be of `Input` component combined with `InputLabel`
- another variation can be on `Input` component with an adornment icon
- and so on ..

This just gives more composition, where we each component can focus on one task at hand, plus this will  make things simpler down the way.

**Comment by peterszerzo** [comment]
I am not sure it will make things simpler - it pushes the responsibility of doing the composition to the user, and allows them to compose combinations that don't make sense from a design perspective, akin to composing an input with a superjumbo icon to illustrate that help is available. The current API makes all the possibilities you listed possible, plus the added rigidity that gives sensible visual guarantees.

**Comment by adeelibr** [comment]
>  it pushes the responsibility of doing the composition to the user, and allows them to compose combinations

But I think that's the entire point of a powerful API that gives the user complete control on how to he/she should compose. If we give them these sort of compositions.

> allows them to compose combinations that don't make sense from a design perspective,

I think that should be left to the user, on how they want to consume it. Limiting just because it doesn't match a certain design spec for a certain use case. Maybe the user has a different design spec they have to follow.

This was just a suggestion. From another perspective 👍

**Comment by peterszerzo** [comment]
Implementing and sticking to a certain design spec is a direction we feel pretty strongly about at the moment. I have always found that seeking flexibility in a UI library is kind of a rabbithole (or were I to put it somewhat crudely: if we wanted complete flexibility, we'd just write raw html and css). And even if it stops at clear boundaries, I'm not sure how good it is for an end product.

**Comment by adeelibr** [comment]
I can understand, your point of view.

**Comment by peterszerzo** [comment]
It looks like we're ready to close up this issue. I did a double-check on the autocomplete attribute's propagation onto the input tag, where this code:

![screen shot 2018-10-04 at 11 06 23](https://user-images.githubusercontent.com/6738398/46464013-9d84a200-c7c5-11e8-8657-25a8c7c91878.png)

renders this html:

![screen shot 2018-10-04 at 11 06 04](https://user-images.githubusercontent.com/6738398/46464038-a5dcdd00-c7c5-11e8-8553-98d327decdc1.png)

so things should be working ok.

@adeelibr I hope the reasoning on keeping inputs and labels glued to each other made sense - it's really a decision we made that we believe sets sensible constraints in order, and not some kind of absolute truth. We appreciate you bringing the point up in any case.
